core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -553,12 +562,14 @@ private[spark] class ExecutorAllocationManager(
         }
 
         // If this is the last pending task, mark the scheduler queue as empty
-        stageIdToTaskIndices.getOrElseUpdate(stageId, new mutable.HashSet[Int]) += taskIndex
+        stageIdToTaskIndices
+          .getOrElseUpdate(stageId, new mutable.HashSet[String]) += (taskIndex + "." + attemptId)
         val numTasksScheduled = stageIdToTaskIndices(stageId).size
         val numTasksTotal = stageIdToNumTasks.getOrElse(stageId, -1)
         if (numTasksScheduled == numTasksTotal) {
           // No more pending tasks for this stage
           stageIdToNumTasks -= stageId
```

Comment:
If anything, I would think that we should remove this line. If this task fails, then the next attempt would go to the else case of `stageIdNumTasks.getOrElse(stageId, -1)`, which not technically correct. It's safe to remove it because we remove it in `stageCompleted` anyway.